### PR TITLE
feat: [CI-8680]: integrated gitness jwt in ci

### PR DESCRIPTION
**Pull request by devkimittal**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/50033)

### CODE REVIEW

File: 879-pipeline-ci-commons/src/main/java/io/harness/ci/utils/BaseConnectorUtils.java [hunk 1]
```diff
@@ -213,7 +212,7 @@ public String getSCMBaseUrl(String baseUrl) {
       String host = url.getHost();
       String protocol = url.getProtocol();
       if (host.equals("localhost")) {
-        return "";
+        return "http://1005-59-89-164-155.ngrok-free.app/git";
```
Comment (raghavharness) [inline 879-pipeline-ci-commons/src/main/java/io/harness/ci/utils/BaseConnectorUtils.java:215]: ?

File: 980-commons/src/main/java/io/harness/security/JWTTokenServiceUtils.java [hunk 2]
```diff
@@ -168,6 +168,31 @@ public String generateJWTToken(Map<String, String> claims, Long validityDuration
     }
   }
 
+  public String generateJWTToken(Map<String, String> claims, Map<String, String[]> arrayClaims, Long validityDurationInMillis, String jwtPasswordSecret) {
+    if (jwtPasswordSecret == null) {
```
Comment (raghavharness) [inline 980-commons/src/main/java/io/harness/security/JWTTokenServiceUtils.java:172]: use isEmpty

Comment (devkimittal) [reply 980-commons/src/main/java/io/harness/security/JWTTokenServiceUtils.java:172]: just copy pasted the already written method. Also secret could be empty as well. So checking null makes sense.

File: 879-pipeline-ci-commons/src/main/java/io/harness/ci/utils/BaseConnectorUtils.java (hunk 1 above)
Comment (devkimittal) [reply 879-pipeline-ci-commons/src/main/java/io/harness/ci/utils/BaseConnectorUtils.java:215]: was testing in local using this url, will remove.

**Comment by devkimittal** [comment]
trigger uts

**Comment by devkimittal** [comment]
trigger uts

**Comment by devkimittal** [comment]
trigger ut0

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/buildstate/CodebaseUtils.java [hunk 3]
```diff
@@ -445,14 +446,20 @@ public ConnectorDetails getGitConnector(NGAccess ngAccess, CodeBase codeBase, bo
     }
 
     String connectorRefValue = codeBase.getConnectorRef().getValue();
-    return getGitConnector(ngAccess, connectorRefValue);
+    String repoName = null;
+    if (codeBase.getRepoName() != null) {
```
Comment (harshjain12) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/buildstate/CodebaseUtils.java:450]: use isNotEmpty

Comment (harshjain12) [reply 332-ci-manager/service/src/main/java/io/harness/ci/execution/buildstate/CodebaseUtils.java:450]: We should use RunTimeInputHandler to resolve parameter field

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/buildstate/ConnectorUtils.java [hunk 4]
```diff
@@ -165,4 +173,37 @@ public ConnectorDetails getConnectorDetails(NGAccess ngAccess, String connectorI
 
     return super.getConnectorDetails(ngAccess, connectorIdentifier);
   }
+
+  public ConnectorDetails getConnectorDetailsWithToken(
+      NGAccess ngAccess, String connectorIdentifier, boolean isGitConnector, Ambiance ambiance, String repoName) {
+    if (isGitConnector && isEmpty(connectorIdentifier)
+        && featureFlagService.isEnabled(FeatureName.CODE_ENABLED, ngAccess.getAccountIdentifier())) {
```
Comment (harshjain12) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/buildstate/ConnectorUtils.java:180]: FF check should be separate if condition

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/buildstate/ConnectorUtils.java [hunk 5]
```diff
@@ -165,4 +173,37 @@ public ConnectorDetails getConnectorDetails(NGAccess ngAccess, String connectorI
...
+  private String fetchAuthToken(NGAccess ngAccess, Ambiance ambiance, String repoName) {
+    ExecutionPrincipalInfo executionPrincipalInfo = ambiance.getMetadata().getPrincipalInfo();
+    String principal = executionPrincipalInfo.getPrincipal();
+    String email = AmbianceUtils.getEmail(ambiance);
+
+    String completeRepoName = GitClientHelper.getCompleteHarnessRepoName(
+        ngAccess.getAccountIdentifier(), ngAccess.getOrgIdentifier(), ngAccess.getProjectIdentifier(), repoName);
+    String[] allowedResources = {completeRepoName};
+
+    ImmutableMap<String, String> claims = ImmutableMap.of("name", principal, "type", "USER", "email", email);
```
Comment (harshjain12) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/buildstate/ConnectorUtils.java:203]: Shall we use constants ?

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/buildstate/ConnectorUtils.java [hunk 6]
```diff
@@ -165,4 +173,37 @@ public ConnectorDetails getConnectorDetails(NGAccess ngAccess, String connectorI
...
+    ImmutableMap<String, String[]> arrayClaims = ImmutableMap.of("allowedResources", allowedResources);
+
+    return JWTTokenServiceUtils.generateJWTToken(claims, arrayClaims, TimeUnit.MILLISECONDS.convert(10, TimeUnit.HOURS),
```
Comment (harshjain12) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/buildstate/ConnectorUtils.java:206]: can we move 10 hr to constant

File: 960-api-services/src/main/java/io/harness/git/GitClientHelper.java [hunk 7]
```diff
@@ -229,6 +231,19 @@ public static String getHarnessApiURL(String url) {
     return getHttpProtocolPrefix(url) + domain;
   }
 
+  public static String getCompleteHarnessRepoName(String accountId, String orgId, String projectId, String repo) {
```
Comment (harshjain12) [inline 960-api-services/src/main/java/io/harness/git/GitClientHelper.java:234]: It seems to be converter

convertToHarnessRepo

**Review by harshjain12 — APPROVED** [review]

**Comment by devkimittal** [comment]
trigger checkstylecodeformat

**Comment by devkimittal** [comment]
trigger uts

**Review by raghavharness — APPROVED** [review]

**Review by harshjain12 — DISMISSED** [review]

**Review by raghavharness — DISMISSED** [review]

**Review by harshjain12 — APPROVED** [review]

**Review by raghavharness — APPROVED** [review]

**Pull request merged**